Tools/ardupilotwaf/ap_libraries.py

```diff
@@ -0,0 +1,140 @@
+necessary task generators for creating the objects of a library for a vehicle.
+That includes the common objects, which are shared among vehicles. That
+function is used by bld.ap_stlib() and shouldn't need to be called otherwise.
+
+The environment variable AP_LIBRARIES_OBJECTS_KW is a dictionary of keyword
+arguments to be passed to bld.objects() when during the creation of the task
+generators. You can use it to pass extra arguments to that function (although
+some of them will be rewritten, see the implementation for details).
+
+NOTE: This tool is provisional, it shouldn't be necessary when all libraries
+become independent of vehicle. Thus, this tool should contain the only
+*minimal* functionality required and addition of features should be avoided.
+"""
+from waflib import Utils
+from waflib.Configure import conf
+from waflib.TaskGen import before_method, feature
+
+import ardupilotwaf as ap
+
+UTILITY_SOURCE_EXTS = ['utility/' + glob for glob in ap.SOURCE_EXTS]
+
+AP_LIBRARIES_VEHICLE_DEPENDENT_SRC = [
+    'libraries/AP_AccelCal/AP_AccelCal.cpp',
+    'libraries/AP_BattMonitor/AP_BattMonitor.cpp',
+    'libraries/AP_Compass/AP_Compass.cpp',
+    'libraries/AP_HAL_Linux/Storage.cpp',
+    'libraries/AP_HAL_PX4/HAL_PX4_Class.cpp',
+    'libraries/AP_HAL_PX4/Storage.cpp',
+    'libraries/AP_HAL_QURT/mainapp/mainapp.cpp',
+    'libraries/AP_HAL_SITL/SITL_cmdline.cpp',
+    'libraries/AP_HAL_VRBRAIN/HAL_VRBRAIN_Class.cpp',
+    'libraries/AP_HAL_VRBRAIN/Storage.cpp',
+    'libraries/AP_InertialSensor/AP_InertialSensor.cpp',
+    'libraries/AP_Mission/AP_Mission.cpp',
+    'libraries/AP_NavEKF2/AP_NavEKF2.cpp',
+    'libraries/AP_NavEKF/AP_NavEKF_core.cpp',
+    'libraries/AP_NavEKF/AP_NavEKF.cpp',
+    'libraries/AP_Rally/AP_Rally.cpp',
+    'libraries/AP_Scheduler/AP_Scheduler.cpp',
+]
+
+_common_objects_tgens = {}
+""" Mapped by library """
+_vehicle_objects_tgens = {}
+""" Mapped by (library, vehicle) """
+
+def _common_tgen_name(library):
+    return 'objs/%s' % library
+
+def _vehicle_tgen_name(library, vehicle):
+    return 'objs/%s/%s' % (library, vehicle)
+
+_vehicle_indexes = {}
+def _vehicle_index(vehicle):
+    """ Used for the objects taskgens idx parameter """
+    if vehicle not in _vehicle_indexes:
+        _vehicle_indexes[vehicle] = len(_vehicle_indexes) + 1
+    return _vehicle_indexes[vehicle]
+
+@conf
+def ap_library(bld, library, vehicle):
+    if library in _common_objects_tgens and \
+       (library, vehicle) in _vehicle_objects_tgens:
+        return
+
+    library_dir = bld.srcnode.find_dir('libraries/%s' % library)
+    if not library_dir:
+        bld.fatal('ap_libraries: %s not found' % library)
+
+    vehicle_dependent = AP_LIBRARIES_VEHICLE_DEPENDENT_SRC
+    sources = library_dir.ant_glob(ap.SOURCE_EXTS + UTILITY_SOURCE_EXTS)
+
+    if library not in _common_objects_tgens:
+        kw = dict(bld.env.AP_LIBRARIES_OBJECTS_KW)
+        kw['features'] = kw.get('features', []) + bld.env.AP_LIBRARY_FEATURES
```

Comment:
What is this AP_LIBRARY_FEATURES for? I mean, don't we pass any feature via AP_LIBRARIES_OBJECTS_KW?